pkg/controller/build/build_controller.go

```diff
@@ -218,3 +244,91 @@ func (r *ReconcileBuild) validateOutputSecret(n string, ns string) error {
 	}
 	return nil
 }
+
+func (r *ReconcileBuild) configFinalizer(b *build.Build) error {
+	if b.Annotations != nil && b.Annotations[build.AnnotationBuildRunDeletion] == "true" {
+		if !contains(b.GetFinalizers(), build.BuildFinalizer) {
+			b.SetFinalizers(append(b.GetFinalizers(), build.BuildFinalizer))
+			err := r.client.Update(context.TODO(), b)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if contains(b.GetFinalizers(), build.BuildFinalizer) {
+			b.SetFinalizers(remove(b.GetFinalizers(), build.BuildFinalizer))
+			err := r.client.Update(context.TODO(), b)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileBuild) finalizeBuildRun(b *build.Build) (bool, error) {
+	// Check if the build is marked to be deleted, which is
+	// indicated by the deletion timestamp being set.
+	if !b.DeletionTimestamp.IsZero() {
+		if contains(b.GetFinalizers(), build.BuildFinalizer) {
+			// Run finalization logic for buildFinalizer. If the
+			// finalization logic fails, don't remove the finalizer so
+			// that we can retry during the next reconciliation.
+			if err := r.cleanBuildRun(b); err != nil {
+				return false, err
+			}
+
+			// Remove buildFinalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			b.SetFinalizers(remove(b.GetFinalizers(), build.BuildFinalizer))
```

Comment:
does this means that even after you called `r.cleanBuildRun` the BuildRun object is not deleted unless you remove the finalizer from the Build??

Reply:
The Build is not deleted.
After all buildRuns are deleted, then the finalizer will be removed,
then the build can be removed.

If the finalizer is still there, the build cannot be removed.

Reply:
got you, thanks